### ISSUE TITLE
Undo Snapshots should be taken after autocorrect and whitespace. And Redo history gets lost. 

### ISSUE THREAD

**Issue by BurntVoxel** [futo-org/android-keyboard#824]
### What happened?

I notice that the undo button doesn't always stop at logical steps. Autocorrect should set a snapshot. 

"I'll type an example that lets you see..."
`I'll type an example that let's [undo]`
`I'll type an example that `|

Huh, it seems likely to get overzealous and undo the entire word, even long ones.
`... get ocerzeakous `|
`... get overzealous `[undo] 
`... get `|

An unrelated bug that I discovered while writing this: redo doesn't seem to function at all? Nothing happened when I wrote a long sentence, hit undo until the sentence was erased, and hit redo repeatedly. I tried both the action button on the top strip and the one under Y. I'll file that bug separately later, have to run somewhere. 

### Keyboard Version

0.1.24-playstore

### Text field info

All textboxes with autocorrect.

**Comment by BurntVoxel** [comment]
Was going to say that it seemed to undo to points in the middle of excessively long words, but seems that did get patched already. Thanks.
... Crap, forgot to change the title.

**Comment by abb128** [comment]
Undo/redo is currently handled by the app, not the keyboard, hence the behavior

**Comment by BurntVoxel** [comment]
Strange. It worked before, but you're right, it's not your keyboard.

Just tried it with Obsidian. Works.
And just tried it with another keyboard on Firefox, doesn't work. ~~(fix yer thing, Mozilla)~~

I take it that implies that the request for autocorrect triggering undo snapshots is kinda impossible unless you build your own Undo mechanic from scratch?

**Comment by abb128** [comment]
Yeah we would have to build our own undo/redo system, at the moment it's not a very high priority. It's not entirely the app's fault when it comes to redo not working because there's two common shortcuts (Ctrl+Y and Ctrl+Shift+Z) but FUTO Keyboard just sends Ctrl+Y. We could try to detect which one works and use that per-app instead of just sending Ctrl+Y every time

**Comment by BurntVoxel** [comment]
It *would* be nice if we could get a keyboard layout that emulates a full QUERTY board, including an actual ctrl key. I like that feature from Unexpected Keyboard.
